runtimes/d8.js

```diff
@@ -38,4 +38,109 @@ var $ = {
   IsHTMLDDA() { return {}; },
   source: $SOURCE,
   realm: Realm.current(),
+  agent: (function() {
+    /*
+      Copyright 2017 the V8 project authors. All rights reserved.
+      Use of this source code is governed by a BSD-style license that can be
+      found in the https://chromium.googlesource.com/v8/v8.git/+/master/LICENSE file.
+      This agent source is adapted from
+      https://chromium.googlesource.com/v8/v8.git/+/master/test/test262/harness-agent.js
+    */
+    var workers = [];
+    var i32a = null;
+    var pendingReports = [];
+
+    // Agents call Atomics.wait on this location to sleep.
+    var SLEEP_LOC = 0;
+    // 1 if the started worker is ready, 0 otherwise.
+    var START_LOC = 1;
+    // The number of workers that have received the broadcast.
+    var BROADCAST_LOC = 2;
+    // Each worker has a count of outstanding reports; worker N uses memory
+    // location [WORKER_REPORT_LOC + N].
+    var WORKER_REPORT_LOC = 3;
+
+    function workerScript(script) {
+      return `
+        var index;
+        var i32a = null;
+        var broadcasts = [];
+        var pendingReceiver = null;
+        function handleBroadcast() {
+          if (pendingReceiver && broadcasts.length > 0) {
+            pendingReceiver.apply(null, broadcasts.shift());
+            pendingReceiver = null;
+          }
+        };
+
+        onmessage = function(msg) {
+          switch (msg.kind) {
+            case 'start':
+              i32a = msg.i32a;
+              index = msg.index;
+              (0, eval)(\`${script}\`);
+              break;
+            case 'broadcast':
+              Atomics.add(i32a, ${BROADCAST_LOC}, 1);
+              broadcasts.push([msg.sab, msg.id]);
+              handleBroadcast();
+              break;
+          }
+        };
+        var $262 = {
+          agent: {
+            receiveBroadcast(receiver) {
+              pendingReceiver = receiver;
+              handleBroadcast();
+            },
+            report(msg) {
+              postMessage(String(msg));
+              Atomics.add(i32a, ${WORKER_REPORT_LOC} + index, 1);
+            },
+            sleep(s) { Atomics.wait(i32a, ${SLEEP_LOC}, 0, s); },
+            leaving() {}
+          }
+        };`;
+    }
+
```

Comment:
@rwaldron - what limitations (if any) does the use of `Atomics` within the agent code here have on being able to cleanly use the v8 agent when running tests for `Atomics.wait` and `Atomics.add` in `test262-harness` ?

Reply:
Discuss this in person with @rwaldron and @jugglinmike - the conclusion was that while this isn't an ideal implementation - we have a uniform API for interacting with agents in test262 and this is an implementation detail which does not affect that API.